### ISSUE TITLE
Epson network printing error recovery

### ISSUE THREAD

**Issue by Syloq** [mike42/escpos-php#57]
I've noticed in testing that once my printer is turned off and eventually turned back on that the next print job sent from the library (after being on) is not sent. However when I send another print job it works fine, but the first print job is never sent and appears the be lost. I thought a possible problem would be the timeout was expiring on the request so instead I send a request (thanks to issue #43) to get the status before every print. I thought that this would help send something to the printer before every print, but it hasn't so far. I'm not sure if this is a library issue or one of the printer as I don't have another printer to verify the functionality.

I have a TM-T88V which I run via the ethernet module. I'm running php 5.5 on Debian.

**Comment by mike42** [comment]
Hi,

Nothing certain, but I've included a few avenues of investigation which might shed light on whether this is a bug.

- I haven't observed this on the TM-T20 (USB) or ethernet-connected Epson printers, which I woudl expect to reject connections until they are ready to print.
 - Check the print server config for any odd settings (log in through the web), and consider re-flashing/updating firmware if it looks old.
- How are you getting the status? ASB should work, but `getPrinterStatus()` is not suggested (it doesn't get processed when the printer is in an error state after the first receipt is printed, which will cause more crashes than it fixes).
 - Remove any image commands and status commands, and see if you can replicate.
- Can you share a snippet of the receipt printing code you're using?
 - I'd be interested to check which `PrintConnector` you're using, if it does anything odd, and what the error handling around it is like (the connectors should throw an `Exception` if they have communications trouble).

**Comment by Syloq** [comment]
I do believe the firmware is updated with the original firmware. So I may have to look into that in the near future.

The status was via the gist in #43 with the ASB. I have removed the ASB call (I have no images being sent) and still get the same result.

My code is a little abstracted, but I can give the barebones of it:

https://gist.github.com/Syloq/ebcdeb55267688aabfb2

Its very possible that my try/catch isn't helping the problem because its hidden inside an ajax request and not being reported correctly for me to see. I can't easily test it right now because of it being used in production, but hopefully I can do that within the next couple weeks.

I'll keep you posted of my findings.

**Comment by Syloq** [comment]
I believe that I have figured this issue out. And I'm not sure there is actually a problem with the library. What I discovered that I couldn't print immediately after the printer is "ready" after being turned on.

On average the printer took about 13 seconds after being turned on to get in a "ready" state and DHCP to serve it an address. What is interesting is that 7 of those seconds are just the printer getting started ready to accept a DHCP packet.

Here was my solution:
```
// getASBStatus() accepts an int that is in seconds for the internal timeout. Defaults to 3.
$status = getASBStatus();
        
 if ($status['status'] === 'offline') {
     for ($step = 1; $step <= 15; $step++) {
         $status = getASBStatus(1);
         if ($status['status'] !== 'offline') {
             break;
         }
     }
 }
```
I realize this may not catch every situation, but enough time should pass that if a ticket is to be printed while the printer is down that there's enough time for the printer to finish its startup sequence. I realize it's not the most elegant, or probably correct, way to do this but this seems to work.
